### PR TITLE
feat(backend): Create SamlConnectionApi and SamlConnection resource

### PR DESCRIPTION
**Pull request by EmmanouelaPothitou**
## Description

Expose any BAPI SAML Connection endpoints as part of every Clerk BE SDK

## Type of change

- [ ] 🐛 Bug fix
- [x] 🌟 New feature
- [ ] 🔨 Breaking change
- [ ] 📖 Refactoring / dependency upgrade / documentation
- [ ] other:

### CODE REVIEW

File: packages/backend/src/api/resources/SamlConnection.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    readonly sp_metadata_url: string,
+    readonly active: boolean,
+    readonly provider: string,
+    readonly user_count: number,
+    readonly sync_user_attributes: boolean,
+    readonly allow_subdomains: boolean,
+    readonly allow_idp_initiated: boolean,
+    readonly created_at: number,
+    readonly updated_at: number,
+    readonly attribute_mapping: AttributeMapping,
```
Comment (nikosdouvlis) [inline packages/backend/src/api/resources/SamlConnection.ts:24]: These class properties need to be camelCased (not snaked_cased). 
In general, the convention we use is to use snake_case for the DTOs (a Data Transfer Object is an object that is used to encapsulate data, and send it from one subsystem of an application to another, eg from our backend to the customer's server using `@clerk/backend) and camelCase within the Clerk package.

To be more precise with an example, even though we use `idp_sso_url` to transfer the sso URL in the network layer, our customers need to be able to access it using:

```ts
const samlConnection = await apiClient.samlConnections.getSamlConnection(...);
samlConnection.idpSsoUrl;
```

File: packages/backend/src/api/endpoints/SamlConnectionApi.ts [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+type UpdateSamlConnectionParams = {
+  name?: string;
+  provider?: SamlIdpSlug;
+  domain?: string;
+  idp_entity_id?: string;
+  idp_sso_url?: string;
+  idp_certificate?: string;
+  idp_metadata_url?: string;
+  attribute_mapping?: {
```
Comment (nikosdouvlis) [inline packages/backend/src/api/endpoints/SamlConnectionApi.ts:37]: Same here, these need to be camelCased. See an example here: https://github.com/clerk/javascript/blob/966b312050b572fbbbc07a6f0581cbec21847375/packages/backend/src/api/endpoints/UserApi.ts#L76

File: packages/backend/src/api/endpoints/SamlConnectionApi.ts [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+type SamlConnectionListParams = {
+  limit?: number;
+  offset?: number;
+};
+type CreateSamlConnectionParams = {
+  name: string;
+  provider: SamlIdpSlug;
+  domain: string;
+  idp_entity_id?: string;
+  idp_sso_url?: string;
```
Comment (nikosdouvlis) [inline packages/backend/src/api/endpoints/SamlConnectionApi.ts:18]: Ditto

File: packages/backend/src/api/endpoints/SamlConnectionApi.ts [hunk 4]
```diff
@@ -0,0 +1,90 @@
+import type { SamlIdpSlug } from '@clerk/types';
+
+import { joinPaths } from '../../util/path';
+import type { SamlConnection } from '../resources';
+import { AbstractAPI } from './AbstractApi';
+
+const basePath = '/saml_connections';
+
+type SamlConnectionListParams = {
```
Comment (chanioxaris) [inline packages/backend/src/api/endpoints/SamlConnectionApi.ts:9]: 🔧 We also support the optional parameters `query` & `orderBy`

File: packages/backend/src/api/endpoints/SamlConnectionApi.ts [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+  offset?: number;
+};
+type CreateSamlConnectionParams = {
+  name: string;
+  provider: SamlIdpSlug;
+  domain: string;
+  idpEntityId?: string;
+  idpSsoUrl?: string;
+  idpCertificate?: string;
+  idpMetadataUrl?: string;
```
Comment (chanioxaris) [inline packages/backend/src/api/endpoints/SamlConnectionApi.ts:20]: 🔧 We are missing the optional `idpMetadata` which is actually the metadata file content

File: packages/backend/src/api/endpoints/SamlConnectionApi.ts [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+};
+
+type UpdateSamlConnectionParams = {
+  name?: string;
+  provider?: SamlIdpSlug;
+  domain?: string;
+  idpEntityId?: string;
+  idpSsoUrl?: string;
+  idpCertificate?: string;
+  idpMetadataUrl?: string;
```
Comment (chanioxaris) [inline packages/backend/src/api/endpoints/SamlConnectionApi.ts:36]: 🔧 Same comment about `idpMetadata` property

File: packages/backend/src/api/resources/AttributeMapping.ts [hunk 7]
```diff
@@ -0,0 +1,14 @@
+import type { AttributeMappingJSON } from './JSON';
```
Comment (chanioxaris) [inline packages/backend/src/api/resources/AttributeMapping.ts:1]: ❓ Maybe we should move this in the saml connection resource file instead as it's part of it and not an different resource. Not sure how we handle these cases

File: packages/backend/src/api/resources/SamlConnection.ts [hunk 8]
```diff
@@ -0,0 +1,49 @@
+import type { AttributeMapping } from './AttributeMapping';
+import type { SamlConnectionJSON } from './JSON';
+
+export class SamlConnection {
```
Comment (chanioxaris) [inline packages/backend/src/api/resources/SamlConnection.ts:4]: 🔧 The following properties can be null `idpEntityId`, `idpSsoUrl`, `idpCertificate`, `idpMetadataUrl`

We also missing the optional `idpMetadata` property

**Review by chanioxaris — COMMENTED** [review]
Great work! Also it seems like there are some commits included not related to our changes here

**Review by nikosdouvlis — APPROVED** [review]
Looks good to me!

File: packages/backend/src/api/resources/SamlConnection.ts [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+    readonly spMetadataUrl: string,
+    readonly active: boolean,
+    readonly provider: string,
+    readonly userCount: number,
+    readonly syncUserAttributes: boolean,
+    readonly allowSubdomains: boolean,
+    readonly allowIdpInitiated: boolean,
+    readonly createdAt: number,
+    readonly updatedAt: number,
+    readonly attributeMapping: AttributeMapping | null,
```
Comment (chanioxaris) [inline packages/backend/src/api/resources/SamlConnection.ts:24]: This property will always be present
```suggestion
    readonly attributeMapping: AttributeMapping,
```

**Review by chanioxaris — APPROVED** [review]
Great work 💯

**Pull request merged**